kale/pipeline/action_domain_adapter.py

```diff
@@ -0,0 +1,674 @@
+# =============================================================================
+# Author: Xianyuan Liu, xianyuan.liu@sheffield.ac.uk
+#         Haiping Lu, h.lu@sheffield.ac.uk or hplu@ieee.org
+# =============================================================================
+
+"""Domain adaptation systems (pipelines) for action video on action recognition"""
+
+import torch
+
+import kale.predict.losses as losses
+from kale.loaddata.video_access import get_image_modality
+from kale.pipeline.domain_adapter import (
+    BaseMMDLike,
+    CDANtrainer,
+    DANNtrainer,
+    get_aggregated_metrics_from_dict,
+    get_metrics_from_parameter_dict,
+    Method,
+    ReverseLayerF,
+    set_requires_grad,
+    WDGRLtrainer,
+)
+
+
+def create_mmd_based_4video(
+    method: Method, dataset, image_modality, feature_extractor, task_classifier, **train_params
+):
+    """MMD-based deep learning methods for action recognition DA: DAN and JAN"""
+    if not method.is_mmd_method():
+        raise ValueError(f"Unsupported MMD method: {method}")
+    if method is Method.DAN:
+        return DANtrainer4Video(
+            dataset=dataset,
+            image_modality=image_modality,
+            feature_extractor=feature_extractor,
+            task_classifier=task_classifier,
+            method=method,
+            **train_params,
+        )
+    if method is Method.JAN:
+        return JANtrainer4Video(
+            dataset=dataset,
+            image_modality=image_modality,
+            feature_extractor=feature_extractor,
+            task_classifier=task_classifier,
+            method=method,
+            kernel_mul=[2.0, 2.0],
+            kernel_num=[5, 1],
+            **train_params,
+        )
+
+
+def create_dann_like_4video(
+    method: Method, dataset, image_modality, feature_extractor, task_classifier, critic, **train_params
+):
+    """DANN-based deep learning methods for action recognition DA: DANN, CDAN, CDAN+E"""
+    # if dataset.is_semi_supervised():
+    #     return create_fewshot_trainer(
+    #         method, dataset, feature_extractor, task_classifier, critic, **train_params
+    #     )
```

Comment:
Flat, comment, or delete?

Reply:
Added more comments to show keeping this for future work.